working/extension_structs/overview.md

```diff
@@ -0,0 +1,543 @@
+
+#### Default constructor
+
+Every non-abstract struct defines a new private constructor with a hidden
+compiler generated name.  We refer to this constructor as the *generated primary
+constructor*.
+
+The generated primary constructor has a single positional parameter for every
+entry in the primary constructor list, each with the obvious type.
+
+If any entry has an initializer value provided, then every entry after (and
+including) that entry is an optional parameter with no default.
+
+For every entry with an initializer value, if no argument is passed for that
+parameter, the initializer value is assigned to that parameter in the
+initializer list of the constructor.  Note that this requires the ability to
+detect whether or not a parameter was passed, which is not expressible strictly
+as a de-sugaring.
+
+If no default constructor is defined in the class, a default constructor is
+generated which forwards to the generated primary constructor.
+
+
+##### Alternative: named parameters
+
+These could be made named parameters.  This feels heavyweight, but has some
+advantages.  The restriction on names in the primary constructor would allow us
+to use the non-private versions of the field names as the parameter names.
+
+#### Equality
+
+If no equality method is defined in or inherited by the struct (except from
+Object), then an equality method will be defined which checks that its argument
+has the same runtime type as the receiver (actual runtime type, not the result
+of calling "runtimeType"), and that the fields of the two objects are pointwise
+equal.
+
+#### hashCode
+
+If no hashCode getter is defined in or inherited by the struct (except from
+Object), then a hashCode getter will be defined which hashes the runtime type
+(again, actual type) together with the hashes of each of the fields.
+
+#### toString
+
+If no toString method is defined in or inherited by the struct (except from
+Object), then a toString method will be defined which returns "struct".
+
+#### debugToString
+
+If no debugToString method is defined in or inherited by the struct, then a
+debugToString method will be defined which returns a formatted description of
+the receiver, of the form `<type>(<f0>, ..., <f1>)` where `<type>` is the
+runtime type of the receiver, and the `fi` are the result of calling
+"debugToString" on the `i`th field if that field is (dynamically) a struct, and
+otherwise the result of calling "toString".
+
+#### copyWith
+
+If no copyWith method is defined in or inherited by the struct, then a copyWith
+method will be defined with named parameters for every entry in the primary
+constructor.  For every entry, if the field name is not private, then the
+parameter name is the field name.  If the field name is private, then the
+parameter name is the field name with all leading `_`s removed.
+
+The body of the method calls the generated primary constructor.  For every
+argument which is explicitly passed to the copyWith method invocation, the
```

Comment:
How should this method determine if an argument is "explicitly passed"?

Reply:
The same way that compilers currently determine this for filling in default values (I believe on the VM via an invocation object, and in dart2js by calling a different entry point).  In other words, compiler magic that users don't get.  I added commentary on this.

Reply:
Ah yeah ok, that makes me sad. I would rather see us make a general user visible feature (its something people have to work around a fair amount today, especially in codegen, and macros will run into it as well).

Reply:
Agreed.

Reply:
Defaulting to an existing field would be something new. Today we can default only to a constant.
If computing the field value is really calling a getter that may have overrides, we might as well allow arbitrary expressions as default values.
I hope there is somewhere between constants and arbitrary expressions.

If we allow `copyWith` to be torn off, the tear-off can be passed to `Function.apply`.
`Function.apply` on dart2js does not use the multitude of entry points, which all match come call site, but a fixed entry point plus metadata.
Changing the defaulting to handle non-constants will make the metadata need to encode receiver-dependent defaults.